File: solutions/python/triangle/1/triangle.py
def equilateral(sides):
    return all(side == sides[0] and side != 0 for side in sides)


def isosceles(sides):
    return is_valid(*sides) and len(set(sides)) in [1, 2]


def scalene(sides):
    return is_valid(*sides) and len(set(sides)) == 3


def is_valid(s1, s2, s3):
    return s1 + s2 >= s3 and s2 + s3 >= s1 and s1 + s3 >= s2

File: solutions/cpp/triangle/1/triangle.cpp
#include <stdexcept>

#include "triangle.h"

namespace triangle {

	flavor kind(double a, double b, double c) {
		if (a <= 0 || b <= 0 || c <= 0
			|| a + b < c || b + c < a || a + c < b)
			throw std::domain_error("invalid triangle");

		if (a == b && a == c && b == c)
			return flavor::equilateral;
		if (a != b && a != c && b != c)
			return flavor::scalene;
		return flavor::isosceles;
	}

}  // namespace triangle
